File: sample/src/main/swift/Sources/SampleAppCore/IntTest.swift
//
// Created by Andrew on 2/4/18.
//

import Foundation

public enum IntEnum: Int, Codable {
    case one
    case two
    case three
}

public struct IntOptionsSet: OptionSet, Codable {

    public let rawValue: Int

    public init(rawValue: Int) {
        self.rawValue = rawValue
    }

    static let one = Int32OptionsSet(rawValue: 1 << 0)
    static let two = Int32OptionsSet(rawValue: 1 << 1)
    static let three = Int32OptionsSet(rawValue: 1 << 2)
}

public struct IntTestStruct: Codable, Hashable {
    public var zero: Int = Int.zero
    public var max: Int = Int(Int32.max)
    public var min: Int = Int(Int32.min)
    public var optional: Int? = Int.zero
    public var optionalNil: Int? = nil
}

public protocol IntTestParamProtocol {
    func testParam(_ param: Int) -> Bool
}

public protocol IntTestReturnTypeProtocol {
    func testReturnType() -> Int
}

public protocol IntTestOptionalParamProtocol {
    func testOptionalParam(_ param: Int?) -> Bool
}

public protocol IntTestOptionalReturnTypeProtocol {
    func testOptionalReturnType() -> Int?
}

public typealias IntBlock = (_ value: Int) -> Int
public typealias OptionalIntBlock = (_ value: Int?) -> Int?

public class IntTest {

    public static func testZero() -> Int {
        return 0
    }

    public static func testMin() -> Int {
        return Int.min
    }

    public static func testMax() -> Int {
        return Int.max
    }

    public static func testMin32() -> Int {
        return Int(Int32.min)
    }

    public static func testMax32() -> Int {
        return Int(Int32.max)
    }

    public static func testParam(_ param: Int) -> Bool {
        return param == Int(Int32.max)
    }

    public static func testReturnType() -> Int {
        return Int(Int32.max)
    }

    public static func testOptionalParam(_ param: Int?) -> Bool {
        return param == Int(Int32.max)
    }

    public static func testOptionalReturnType() -> Int? {
        return Int.max
    }

    public static func testOptional32ReturnType() -> Int? {
        return Int(Int32.max)
    }

    public static func testProtocolParam(_ callback: IntTestParamProtocol) -> Bool {
        return callback.testParam(Int(Int32.max))
    }

    public static func testProtocolReturnType(_ callback: IntTestReturnTypeProtocol) -> Int {
        return callback.testReturnType()
    }

    public static func testProtocolOptionalParam(_ callback: IntTestOptionalParamProtocol) -> Bool {
        return callback.testOptionalParam(Int(Int32.max))
    }

    public static func testProtocolOptionalReturnType(_ callback: IntTestOptionalReturnTypeProtocol) -> Int? {
        return callback.testOptionalReturnType()
    }

    public static func testEncode() ->  IntTestStruct {
        return IntTestStruct()
    }

    public static func testDecode(_ value: IntTestStruct) -> Bool {
        return value == IntTestStruct()
    }

    public static func testEnumEncode(_ rawValue: Int) -> IntEnum {
        switch rawValue {
        case IntEnum.one.rawValue: return IntEnum.one
        case IntEnum.two.rawValue: return IntEnum.two
        case IntEnum.three.rawValue: return IntEnum.three
        default: fatalError("Can't find enum with rawValue \(rawValue)")
        }
    }

    public static func testEnumDecode(_ enumValue: IntEnum) -> Int {
        return enumValue.rawValue
    }

    public static func testOptionSetEncode(_ rawValue: Int) -> IntOptionsSet {
        return IntOptionsSet(rawValue: rawValue)
    }

    public static func testOptionSetDecode(_ optionSet: IntOptionsSet) -> Int {
        return optionSet.rawValue
    }

    public static func testBlock(_ block: IntBlock) -> Bool {
        let value = block(0)
        return value == 0
    }

    public static func testOptionalBlock(_ block: OptionalIntBlock) -> Bool {
        let value = block(nil)
        return value == nil
    }

}

File: sample/src/main/java/com/readdle/swiftjava/sample/IntTest.kt
package com.readdle.swiftjava.sample

import com.readdle.codegen.anotation.*
import java.lang.annotation.Native

@SwiftValue
enum class IntEnum(val rawValue: Int) {

    ONE(0), TWO(1), THREE(2);

    companion object {

        private val values = HashMap<Int, IntEnum>()

        @JvmStatic
        fun valueOf(rawValue: Int): IntEnum {
            return values[rawValue]!!
        }

        init {
            enumValues<IntEnum>().forEach {
                values[it.rawValue] = it
            }
        }
    }

}

@SwiftValue
data class IntOptionsSet(val rawValue: Int = 0) {
    companion object {
        @JvmStatic
        val  one = IntOptionsSet(1)
        @JvmStatic
        val  two = IntOptionsSet(2)
        @JvmStatic
        val  three = IntOptionsSet(4)
    }
}

@SwiftValue
data class IntTestStruct(var zero: Int = 0,
                         var max: Int = Int.MAX_VALUE,
                         var min: Int = Int.MIN_VALUE,
                         var optional: Int? = 0,
                         var optionalNil: Int? = null)

@SwiftReference
class IntTest private constructor() {

    @SwiftDelegate(protocols = ["IntTestParamProtocol"])
    interface IntParamProtocol {
        @SwiftCallbackFunc
        fun testParam(param: Int): Boolean
    }

    @SwiftDelegate(protocols = ["IntTestReturnTypeProtocol"])
    interface IntReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testReturnType(): Int
    }

    @SwiftDelegate(protocols = ["IntTestOptionalParamProtocol"])
    interface IntOptionalParamProtocol {
        @SwiftCallbackFunc
        fun testOptionalParam(param: Int?): Boolean
    }

    @SwiftDelegate(protocols = ["IntTestOptionalReturnTypeProtocol"])
    interface IntOptionalReturnTypeProtocol {
        @SwiftCallbackFunc
        fun testOptionalReturnType(): Int?
    }

    @SwiftBlock("(Int) -> Int")
    interface IntBlock {
        fun call(value: Int): Int
    }

    @SwiftBlock("(Int?) -> Int?")
    interface OptionalIntBlock {
        fun call(value: Int?): Int?
    }

    companion object {
        @JvmStatic
        external fun testZero(): Int

        @JvmStatic
        external fun testMin(): Int

        @JvmStatic
        external fun testMax(): Int

        @JvmStatic
        external fun testParam(param: Int): Boolean

        @JvmStatic
        external fun testReturnType(): Int

        @JvmStatic
        external fun testOptionalParam(param: Int?): Boolean

        @JvmStatic
        external fun testOptionalReturnType(): Int?

        @JvmStatic
        external fun testProtocolParam(callback: IntParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolReturnType(callback: IntReturnTypeProtocol): Int

        @JvmStatic
        external fun testProtocolOptionalParam(callback: IntOptionalParamProtocol): Boolean

        @JvmStatic
        external fun testProtocolOptionalReturnType(callback: IntOptionalReturnTypeProtocol): Int?

        @JvmStatic
        external fun testEncode(): IntTestStruct

        @JvmStatic
        external fun testDecode(value: IntTestStruct): Boolean

        @JvmStatic
        external fun testEnumEncode(rawValue: Int) : IntEnum

        @JvmStatic
        external fun testEnumDecode(enum: IntEnum) : Int

        @JvmStatic
        external fun testOptionSetEncode(rawValue: Int) : IntOptionsSet

        @JvmStatic
        external fun testOptionSetDecode(enum: IntOptionsSet) : Int
        
        @JvmStatic
        external fun testBlock(@SwiftBlock block: IntBlock): Boolean

        @JvmStatic
        external fun testOptionalBlock(@SwiftBlock block: OptionalIntBlock): Boolean
    }

    @Native
    var nativePointer: Long = 0

    external fun release()

}